### DISCUSSION TITLE
can i use DuckDBClient in node.js data loader?

### DISCUSSION THREAD

**Discussion by chen56** [observablehq/framework#1721] [Q&A] [answered]
Can I use DuckDBClient in the data loader? If so, is there an example? I found that the existing example (observablehq/framework/examples/loader-duckdb) is shell-based. But I want to use js + DuckDBClient. Is this okay?

**Comment by mbostock** [comment, marked as answer]
Framework’s `DuckDBClient` client is currently built-in to Framework (it’s `observablehq:stdlib/duck`), so you can’t import it into a Node.js program. There’s an open issue #852 to publish a few of these built-in libraries to npm. You might be able to copy [the source code](https://github.com/observablehq/framework/blob/main/src/client/stdlib/duckdb.js) and with a few changes get it working in Node.js.

That said, it’s probably better to use the official [Node.js API for DuckDB](https://duckdb.org/docs/api/nodejs/overview.html) in a Node.js data loader rather than using Framework’s `DuckDBClient`. The [`duckdb-async` package](https://www.npmjs.com/package/duckdb-async) provides a Promise-based API (and TypeScript types, if you want to use a TypeScript data loader).

**Reply by chen56** [reply]
Thanks for your quick reply!
